### ISSUE TITLE
Gap between X-axis and and imshow 2D plot in matplotlib 2.0.0

### ISSUE THREAD

**Issue by fkbreitl** [matplotlib/matplotlib#8010]
These lines produce a gap between the X axis and the image:

```python
    plt.rcParams['figure.figsize'] = [8, 7]
    plt.imshow([[1, 3], [2, 1]])
    plt.savefig('gap.jpg', dpi=50)
```

![gap](https://cloud.githubusercontent.com/assets/1030293/22567616/eda69d8e-e990-11e6-8740-8222301690de.jpg)

**Comment by dstansby** [comment]
Yep, this is reproducible on current master. It is saving to a .jpg file with a low dpi that causes the gap to appear.

**Comment by fkbreitl** [comment]
The cause is not just the low dpi but more complex.
If I use `plt.rcParams['figure.figsize'] = [8, 7.01]` or `dpi=49` the gab disappears.

**Comment by WeatherGod** [comment]
This is probably a snapping issue, much akin to the issues we have had with
lines appearing in colorbars. This is likely to be extremely difficult to
solve.

**Comment by fkbreitl** [comment]
I can imagine that it could be hard to fix but also a good thing to do since its an indication of something not running as smoothly as it should which could cause all sorts of trouble later.

**Comment by QuLogic** [comment]
Seems to still be a problem on `main`, and not specific to `jpg`:
![gap](https://user-images.githubusercontent.com/302469/234485230-9816046e-4ec9-4d3d-ad95-60f58c9e912a.png)
